File: shop-api/src/main/java/com/javuar/shop/common/sort/SortBy.java
package com.javuar.shop.common.sort;

import lombok.experimental.UtilityClass;

@UtilityClass
public class SortBy {

    private enum SortByProduct implements SortableProperty {
        CREATED_DATE("createdDate"),
        NAME("name"),
        RATE("rate"),
        POPULARITY("popularity"),
        AVAILABILITY_ASC("availabilityAsc"),
        AVAILABILITY_DESC("availabilityDesc"),
        PRICE_ASC("priceAsc"),
        PRICE_DESC("priceDesc");

        private final String property;

        SortByProduct(String property) {
            this.property = property;
        }

        @Override
        public String getProperty() {
            return this.property;
        }
    }

    private enum SortByFeedback implements SortableProperty {
        NOTE("note");

        private final String property;

        SortByFeedback(String property) {
            this.property = property;
        }

        @Override
        public String getProperty() {
            return this.property;
        }
    }

    // Returns full enum class based on domain name
    public Class<? extends Enum<?>> getForDomain(String domain) {
        return switch (domain) {
            case "product" -> SortByProduct.class;
            case "feedback" -> SortByFeedback.class;
            default -> null;
        };
    }
}

File: shop-ui/src/app/pages/components/product-search-options/enums/sort-by.ts
export enum SortBy {
  NAME = 'name',
  CREATED_DATE = 'createdDate',
  RATE = 'rate',
  POPULARITY = 'popularity',
  AVAILABILITY = 'availability',
  PRICE = 'price'
}

export const SORT_BY2 = ['name', 'createdDate', 'rate', 'popularity', 'availability', 'price'] as const;
export type SortBy2 = typeof SORT_BY2[number];

export const SortBy3 = {
  NAME: 'name',
  CREATED_DATE: 'createdDate',
  RATE: 'rate',
  POPULARITY: 'popularity',
  AVAILABILITY: 'availability',
  PRICE: 'price'
} as const

export type SortBy3 = typeof SortBy3[keyof typeof SortBy];
